#include "addinputdialog.h"

//#include "qpushbutton.h"
#include "model/cutting/plan/request.h"
#include "ui_addinputdialog.h"
#include "model/registries/materialregistry.h"

#include <QMessageBox>

AddInputDialog::AddInputDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::AddInputDialog)
    , current_requestId(QUuid::createUuid()) // ðŸ”‘ Automatikusan Ãºj UUID
{
    ui->setupUi(this);
    populateMaterialCombo();

    // ui->comboCategory->addItem("RollerTube");
    // ui->comboCategory->addItem("BottomBar");

    // // Ok gomb disable ha nincs Ã©rtelmes adat
    // connect(ui->editLength, &QLineEdit::textChanged, this, [=]() {
    //     ui->buttonBox->button(QDialogButtonBox::Ok)
    //         ->setEnabled(!ui->editLength->text().isEmpty());
    // });
}

AddInputDialog::~AddInputDialog()
{
    delete ui;
}


void AddInputDialog::populateMaterialCombo() {
    const auto& registry = MaterialRegistry::instance().readAll();

    ui->comboMaterial->clear();
    for (const auto& m : registry) {
        ui->comboMaterial->addItem(m.toDisplay(), m.id);
    }
}

QUuid AddInputDialog::selectedMaterialId() const {
    return ui->comboMaterial->currentData().toUuid();
}

int AddInputDialog::length() const
{
    return ui->editLength->text().toInt();
}

int AddInputDialog::quantity() const
{
    return ui->spinQuantity->value();
}

QString AddInputDialog::ownerName() const {
    return ui->editOwner->text().trimmed();
}

QString AddInputDialog::externalReference() const {
    return ui->editReference->text().trimmed();
}

Cutting::Plan::Request AddInputDialog::getModel() const {
    Cutting::Plan::Request req;
    req.requestId = current_requestId; // âœ… ez volt a hiÃ¡nyzÃ³ lÃ¡ncszem
    // ðŸ”— Anyag ID kinyerÃ©se a comboBox-bÃ³l
    QVariant matData = ui->comboMaterial->currentData();
    if (matData.isValid())
        req.materialId = matData.toUuid();

    // ðŸ“ VÃ¡gÃ¡si hossz kiolvasÃ¡sa
    bool okLen = false;
    req.requiredLength = ui->editLength->text().toInt(&okLen);
    if (!okLen)
        req.requiredLength = -1; // HibÃ¡s hossz

    // ðŸ”¢ DarabszÃ¡m
    req.quantity = quantity(); // <- ha mÃ¡r van quantity() metÃ³dusod

    // ðŸ‘¤ MegrendelÅ‘ neve
    req.ownerName = ownerName(); // <- ha mÃ¡r van ownerName() getter

    // ðŸ§¾ KÃ¼lsÅ‘ tÃ©telszÃ¡m
    req.externalReference = externalReference(); // <- ha van ilyen getter

    return req;
}


bool AddInputDialog::validateInputs() {
    Cutting::Plan::Request req = getModel(); // <- Ãºj metÃ³dusod, lÃ¡sd korÃ¡bban

    QStringList errors = req.invalidReasons(); // <- centralizÃ¡lt validÃ¡ciÃ³

    if (!errors.isEmpty()) {
        QMessageBox::warning(this,
                             "AdatellenÅ‘rzÃ©s",
                             "KÃ©rlek javÃ­tsd az alÃ¡bbi hibÃ¡kat:\n\n" + errors.join("\n"));
        return false;
    }

    return true;
}


void AddInputDialog::accept() {
    if (!validateInputs())
        return;

    QDialog::accept(); // csak ha minden okÃ©
}

void AddInputDialog::setModel(const Cutting::Plan::Request& request) {
    current_requestId = request.requestId; // â¬…ï¸ ID mentÃ©se

    // ðŸ”— Anyag beÃ¡llÃ­tÃ¡sa comboBox-ban
    int index = ui->comboMaterial->findData(request.materialId);
    if (index >= 0)
        ui->comboMaterial->setCurrentIndex(index);

    // ðŸ“ Hossz
    ui->editLength->setText(QString::number(request.requiredLength));

    // ðŸ”¢ DarabszÃ¡m
    ui->spinQuantity->setValue(request.quantity);

    // ðŸ‘¤ MegrendelÅ‘
    ui->editOwner->setText(request.ownerName);

    // ðŸ§¾ KÃ¼lsÅ‘ azonosÃ­tÃ³
    ui->editReference->setText(request.externalReference);
}


